test/core.jl

```diff
@@ -704,6 +704,27 @@ end
 @test foo21900 == 10
 @test bar21900 == 11
 
+let f = g -> x -> g(x)
+    @test f(Int)(1.0) === 1
+    @test @inferred(f(Int)) isa Function
+    @test fieldtype(typeof(f(Int)), 1) === Type{Int}
+    @test @inferred(f(Rational)) isa Function
+    @test fieldtype(typeof(f(Rational)), 1) === Type{Rational}
+    @test_broken @inferred(f(Rational{Core.TypeVar(:T)})) isa Function
+    @test fieldtype(typeof(f(Rational{Core.TypeVar(:T)})), 1) === DataType
+end
+let f() = (T = Rational{Core.TypeVar(:T)}; () -> T)
+    t = f()()
+    @test t isa DataType
+    @test t.name.wrapper == Rational
+    @test length(t.parameters) == 1
+    @test t.parameters[1] isa Core.TypeVar
+
+    @test_broken @inferred(f()) isa Function
```

Comment:
Any ideas what's going on here?
```julia
julia> g() = (T = Rational{Core.TypeVar(:T)}; () -> T)
g (generic function with 1 method)

julia> ir = @code_typed g()
CodeInfo(
1 ─ %1  = Core._typevar(:T, Union{}, Any)::TypeVar
│   %2  = Core.apply_type(Main.Rational, %1)::Type{Rational{T}}
│   %3  = Main.:(var"#3#4")::Type{var"#3#4"}
│   %4  = Core.has_free_typevars::typeof(Core.has_free_typevars)
│   %5  = (%4)(%2)::Bool
└──       goto #3 if not %5
2 ─ %7  = Base.typeof(%2)::DataType
└──       goto #4
3 ─ %9  = Core.apply_type(Core.Type, %2)::Type{Type{_A}} where _A
└──       goto #4
4 ┄ %11 = φ (#2 => %7, #3 => %9)::DataType
│   %12 = Core.apply_type(%3, %11)::Type{var"#3#4"{_A}} where _A
│   %13 = %new(%12, %2)::var"#3#4"
└──       return %13
) => var"#3#4"

julia> Core.Compiler.has_free_typevars_tfunc(ir[1].ssavaluetypes[2])
Core.Const(true)
```
Why is `%5` inferred as `Bool` instead of `Core.Const(true)`?

Reply:
You need to inspect unoptimized frame to see how abstract interpretation goes. Essentially when inference enters the frame of `_typeof_captured_variable` we intentionally widen anything including free typevars (as free typevars are very annoying): https://github.com/JuliaLang/julia/blob/2ba3a22858938ecaf41bbe3e780df507556825c5/base/compiler/inferenceresult.jl#L168

I think the easiest direction here would be to have a special case in `abstract_known_call` rather than having a new builtin.

Reply:
#43965

Reply:
> we intentionally widen anything including free typevars

So, I guess this is why we can't just mark `has_free_typevars` as pure and expect to get a similar result?

Reply:
If we need to forward it to `_typeof_captured_variable`, yes. If the frontend inserts (inlines) the entire body of `_typeof_captured_variable`, then we don't need to forward it and it would probably work to just mark `has_free_typevars` as pure.

Reply:
Marking `_typeof_captured_variable` as pure is also not enough?

Reply:
Well, pure function call can be folded only when all the inputs are `Const`. But we want to infer `_typeof_captured_variable(::Int)` as `Type{Int}` for example and it requires some special casing anyway.

Reply:
What about

```julia
_typeof_captured_variable(::T) where {T} = T
@pure _typeof_captured_variable(::Type{T}) where {T} = has_free_typevars(T) ? typeof(T) : Typeof(T)
```

? i.e., Can we extract out special casing from the compiler?

Reply:
I believe that would run into the same issues I was having above that free type vars are widened automatically.